server/src/internalClusterTest/java/org/elasticsearch/index/engine/CheckAbortedMergesIT.java

```diff
@@ -0,0 +1,364 @@
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.stream.StreamSupport;
+
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertNoFailures;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.greaterThan;
+
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST, numDataNodes = 0)
+public class CheckAbortedMergesIT extends ESIntegTestCase {
+
+    @Override
+    protected boolean addMockInternalEngine() {
+        return false;
+    }
+
+    @Override
+    protected Collection<Class<? extends Plugin>> nodePlugins() {
+        var plugins = new ArrayList<>(super.nodePlugins());
+        plugins.add(BlockRunningMergesEngineTestPlugin.class);
+        return plugins;
+    }
+
+    public void testAbortedMerges() throws Exception {
+        internalCluster().startMasterOnlyNode();
+        var nodeA = internalCluster().startDataOnlyNode();
+
+        var pluginA = internalCluster().getInstance(PluginsService.class, nodeA)
+            .filterPlugins(BlockRunningMergesEngineTestPlugin.class)
+            .findFirst()
+            .orElseThrow(() -> new AssertionError("Plugin not found"));
+
+        final boolean checkAbortedMerges = randomBoolean();
+        pluginA.blockMerges();
+
+        final var indexName = randomIdentifier();
+        createIndex(
+            indexName,
+            indexSettings(1, 0).put(CheckAbortedDuringMergePolicy.ENABLE_CHECK_ABORTED_DURING_MERGE.getKey(), checkAbortedMerges)
+                .put(IndexSettings.INDEX_REFRESH_INTERVAL_SETTING.getKey(), TimeValue.MINUS_ONE)
+                .build()
+        );
+
+        var indexServiceA = internalCluster().getInstance(IndicesService.class, nodeA).indexService(resolveIndex(indexName));
+        assertThat(indexServiceA.hasShard(0), equalTo(true));
+
+        indexDocs(indexName, 10);
+        flush(indexName);
+
+        while (true) {
+            indexDocs(indexName, 10);
+            flush(indexName);
+
+            var mergesStats = client().admin().indices().prepareStats(indexName).clear().setMerge(true).get();
+            if (mergesStats.getIndices().get(indexName).getPrimaries().getMerge().getCurrent() > 0) {
+                break;
+            }
+        }
+
+        var nodeB = internalCluster().startDataOnlyNode();
+        ensureStableCluster(3);
+
+        pluginA.waitForMergesBlocked();
+
+        ClusterRerouteUtils.reroute(client(), new MoveAllocationCommand(indexName, 0, nodeA, nodeB, ProjectId.DEFAULT));
+        ensureGreen(indexName);
+
+        var indexServiceB = internalCluster().getInstance(IndicesService.class, nodeB).indexService(resolveIndex(indexName));
+        assertBusy(() -> assertThat(indexServiceB.hasShard(0), equalTo(true)));
```

Comment:
nit: isn't this implicitly asserted by `ensureGreen`? it waits for relocations to complete, but maybe it's good to have the explicit assertion.